ports/zephyr/machine_spi.c

```diff
@@ -0,0 +1,217 @@
+        { MP_QSTR_phase, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_PHASE} },
+        { MP_QSTR_bits, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_BITS} },
+        { MP_QSTR_firstbit, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_FIRSTBIT} }, 
+        { MP_QSTR_sck, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_mosi, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_miso, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    const char *dev_name = mp_obj_str_get_str(args[ARG_id].u_obj);
+    const struct device *dev = device_get_binding(dev_name);
+
+    if (dev == NULL) {
+        mp_raise_ValueError(MP_ERROR_TEXT("device not found"));
+    }
+
+    if ((args[ARG_sck].u_obj != MP_OBJ_NULL) || (args[ARG_miso].u_obj != MP_OBJ_NULL) || (args[ARG_mosi].u_obj != MP_OBJ_NULL)) {
+        mp_raise_NotImplementedError(MP_ERROR_TEXT("explicit choice of sck/miso/mosi is not implemented"));
+    }
+
+    struct spi_config cfg = {
+        .frequency = args[ARG_baudrate].u_int,
+        .operation = SPI_OP_MODE_MASTER |
+                args[ARG_polarity].u_int << 1 |
+                args[ARG_phase].u_int << 2 |
+                SPI_LOOP << 3 |
+                args[ARG_firstbit].u_int << 4 |
+                args[ARG_bits].u_int << 5 |
+                SPI_LINES_SINGLE,
+        .slave = 0,
+        .cs = NULL
+    };
+
+    machine_hard_spi_obj_t *self = m_new_obj(machine_hard_spi_obj_t);
+
+    self->base.type = &machine_hard_spi_type;
+    self->dev = dev;
+    self->config = cfg;
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+
+STATIC void machine_hard_spi_init(mp_obj_base_t *obj, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+
+    enum {ARG_baudrate, ARG_polarity, ARG_phase, ARG_bits, ARG_firstbit};
+
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_baudrate, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_polarity, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_phase, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_bits, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_firstbit, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    machine_hard_spi_obj_t *self = (machine_hard_spi_obj_t*)obj;
+
+    uint32_t baudrate;
+    uint16_t operation = self->config.operation;
+
+    if (args[ARG_baudrate].u_int != -1) {
+        baudrate = args[ARG_baudrate].u_int;
+    } else {
+        baudrate = self->config.frequency;
+    }
+
+    if (args[ARG_polarity].u_int != -1) {
+        operation = (operation & ~0x2) | (args[ARG_polarity].u_int << 1);
```

Comment:
Shouldn't those args values be range checked to make sure they do not overflow the field and overwrite other parameters?

Reply:
Ports can range check values at their discretion (eg if there's not much flash then range checking will be omitted to save code size).  It's generally undefined behaviour if the user (accidentally) passes in the wrong value, for example with SPI's firstbit argument, it must be SPI.LSB or SPI.MSB.